### PR TITLE
Add EMR client to spark connector

### PR DESCRIPTION
**Pull request by rupal-bq**
### Description
- Add EMR client with sync approach mentioned in https://github.com/opensearch-project/sql/issues/1721
- This is follow up for https://github.com/opensearch-project/sql/pull/1780

DO NOT MERGE before https://github.com/opensearch-project/sql/pull/1780
 
### Issues Resolved
- https://github.com/opensearch-project/sql/issues/1721
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/storage/SparkMetricScan.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.opensearch.sql.spark.request.SparkQueryRequest;
+import org.opensearch.sql.spark.response.SparkResponse;
+import org.opensearch.sql.storage.TableScanOperator;
+
+/**
+ * Spark scan operator.
+ */
+@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
+@ToString(onlyExplicitlyIncluded = true)
+public class SparkMetricScan extends TableScanOperator {
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/storage/SparkMetricScan.java:30]: SparkMetricScan, could you find better name?

File: spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  @Override
+  public void close() {
+    super.close();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return this.sparkResponseHandle.hasNext();
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java:57]: does SparkSqlFunctionResponseHandle required, it do same thing as SparkSqlFunctionTableScanOperator

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  public EmrClientImpl(
+      Client client, EMRHelper emr, FlintHelper flint) {
+    this.client = client;
+    this.emr = emr;
+    this.flint = flint;
+  }
+
+  @Override
+  public JSONObject sql(String query) throws IOException {
+    return new SparkResponse(
```
Comment (penghuo) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:54]: Is SparkResponse helper class?
If yes, make it static.

**Comment by penghuo** [comment]
when do we delete the query result?

File: core/src/main/java/org/opensearch/sql/data/model/ExprBooleanValue.java [hunk 4]
```diff
@@ -19,7 +19,7 @@ public class ExprBooleanValue extends AbstractExprValue {
 
   private final Boolean value;
 
-  private ExprBooleanValue(Boolean value) {
+  public ExprBooleanValue(Boolean value) {
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply core/src/main/java/org/opensearch/sql/data/model/ExprBooleanValue.java]: Yes, Thanks for catching this. used of().

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  private void waitForStepExecution(DescribeStepRequest stepRequest) {
+    // Wait for the step to complete
+    boolean completed = false;
+    while (!completed) {
+      // Get the step status
+      StepStatus statusDetail = emr.getStepStatus(stepRequest);
+      // Check if the step has completed
+      if (statusDetail.getState().equals("COMPLETED")) {
+        completed = true;
+        logger.info("EMR step completed successfully.");
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:110]: application id is added in log on line 87 above.

File: spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.spark.data.constants;
+
+public class SparkConstants {
+  public static final String EMR = "emr";
+  public static final String STEP_ID_FIELD = "stepId.keyword";
+  public static final String SPARK_APPLICATION_JAR = "s3://spark-datasource/sql-job.jar";
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/data/constants/SparkConstants.java]: added config option in properties.

File: spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        result.add(new ExprTupleValue(linkedHashMap));
+      }
+    } else {
+      throw new RuntimeException("Unexpected result during spark sql query execution");
+    }
+    this.schema = new ExecutionEngine.Schema(columnList);
+    this.responseIterator = result.iterator();
+  }
+
+  @NotNull
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java]: Thanks for catching this. not needed. removed.

File: spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+import org.opensearch.sql.data.type.ExprCoreType;
+import org.opensearch.sql.data.type.ExprType;
+import org.opensearch.sql.executor.ExecutionEngine;
+
+/**
+ * Default implementation of SparkSqlFunctionResponseHandle.
+ */
+public class DefaultSparkSqlFunctionResponseHandle implements SparkSqlFunctionResponseHandle {
+
+  private final JSONObject responseObject;
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java]: No. removed.

File: spark/src/main/java/org/opensearch/sql/spark/storage/SparkMetricScan.java (hunk 1 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/storage/SparkMetricScan.java]: renamed to SparkScan

**Comment by rupal-bq** [comment]
> when do we delete the query result?

We need to set ISM policy for result index for deleting results.

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    String stepId = result.getStepIds().get(0);
+    DescribeStepRequest stepRequest = new DescribeStepRequest()
+        .withClusterId(emr.getEmrCluster())
+        .withStepId(stepId);
+
+    waitForStepExecution(stepRequest);
+    sparkResponse.setValue(stepId);
+
+    return stepId;
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:97]: The return value is not used. maybe return sparkResponse instead?

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    StepConfig emrstep = new StepConfig()
+        .withName("Spark Application")
+        .withActionOnFailure(ActionOnFailure.CONTINUE)
+        .withHadoopJarStep(stepConfig);
+
+    AddJobFlowStepsRequest request = new AddJobFlowStepsRequest()
+        .withJobFlowId(emr.getEmrCluster())
+        .withSteps(emrstep);
+
+    AddJobFlowStepsResult result = emr.addStep(request);
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:86]: trying to figure out which line starts the job actually. I'm thinking we may need to add metrics?

File: spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            LOG.error(e.getMessage());
+            throw new RuntimeException(
+                String.format("Error fetching data from spark server: %s", e.getMessage()));
+          }
+        });
+  }
+
+  @Override
+  public void close() {
+    super.close();
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java:52]: just to confirm no resource to release from our plugin side?

File: spark/src/main/java/org/opensearch/sql/spark/helper/FlintHelper.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DEFAULT_AUTH;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DEFAULT_HOST;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DEFAULT_PORT;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DEFAULT_REGION;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DEFAULT_SCHEME;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INTEGRATION_JAR;
+
+import lombok.Getter;
+
+public class FlintHelper {
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/helper/FlintHelper.java:17]: This is actually FlintConfig?

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      if (statusDetail.getState().equals("COMPLETED")) {
+        completed = true;
+        logger.info("EMR step completed successfully.");
+      } else if (statusDetail.getState().equals("FAILED")
+          || statusDetail.getState().equals("CANCELLED")) {
+        logger.error("EMR step failed or cancelled.");
+        throw new RuntimeException("Spark SQL application failed.");
+      } else {
+        // Sleep for some time before checking the status again
+        Thread.sleep(2500);
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:117]: just wonder is this arbitrary or this is the average latency per check?

File: spark/src/main/java/org/opensearch/sql/spark/storage/implementor/SparkDefaultImplementor.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.RequiredArgsConstructor;
+import org.opensearch.sql.planner.DefaultImplementor;
+import org.opensearch.sql.spark.storage.SparkScan;
+
+/**
+ * Default Implementor of Logical plan for spark.
+ */
+@RequiredArgsConstructor
+public class SparkDefaultImplementor
+    extends DefaultImplementor<SparkScan> {
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/storage/implementor/SparkDefaultImplementor.java:17]: If no override logic, this subclass is not necessary?

File: spark/src/main/java/org/opensearch/sql/spark/storage/SparkScan.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import org.opensearch.sql.spark.client.SparkClient;
+import org.opensearch.sql.spark.request.SparkQueryRequest;
+import org.opensearch.sql.storage.TableScanOperator;
+
+/**
+ * Spark scan operator.
+ */
+@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
+@ToString(onlyExplicitlyIncluded = true)
+public class SparkScan extends TableScanOperator {
```
Comment (dai-chen) [inline spark/src/main/java/org/opensearch/sql/spark/storage/SparkScan.java:22]: What's the difference between this class and SparkSqlFunctionTableScanOperator?

File: spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+      JSONObject items = responseObject.getJSONObject("data");
+      columnList = getColumnList(items.getJSONArray("schema"));
+      for (int i = 0; i < items.getJSONArray("result").length(); i++) {
+        JSONObject row = new JSONObject(
+            items.getJSONArray("result").get(i).toString().replace("'", "\""));
+        LinkedHashMap<String, ExprValue> linkedHashMap = extractRow(row, columnList);
+        result.add(new ExprTupleValue(linkedHashMap));
+      }
+    } else {
+      throw new RuntimeException("Unexpected result during spark sql query execution");
```
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/response/DefaultSparkSqlFunctionResponseHandle.java:59]: No. it should still have stepId & applicationId information. This check was just precautionary in case something went wrong with json object. can be removed. 
Also tested create query and received empty response. 
```
{
    "schema": [],
    "datarows": [],
    "total": 0,
    "size": 0
}
```

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java (hunk 13 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:116]: just arbitrary value

File: spark/src/main/java/org/opensearch/sql/spark/helper/FlintHelper.java (hunk 12 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/helper/FlintHelper.java:17]: yes this is for passing required parameters to flint integration jar

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java (hunk 3 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java]: setting stepId afterwards so can't make it static.

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java (hunk 9 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java]: thanks! removed return value.

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java (hunk 10 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java]: `addJobFlowSteps` starts the job(emr step). Didn't cover metrics in design. will take it up as follow up task.

File: spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java (hunk 11 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java]: yes. removed this. We need to delete results from index afterwards but need to check ISM policy for it.

File: spark/src/main/java/org/opensearch/sql/spark/storage/implementor/SparkDefaultImplementor.java (hunk 14 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/storage/implementor/SparkDefaultImplementor.java]: Thanks for catching this. removed subclass.

File: spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/client/EmrClientImpl.java:109]: done

File: spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java (hunk 2 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java:52]: Not required. Can be removed. I tried to follow similar structure as prometheus as it seemed good abstraction to me.

**Comment by rupal-bq** [comment]
> when do we delete the query result?

Added delete query result after retrieving from index `.query_execution_result`

File: spark/src/main/java/org/opensearch/sql/spark/storage/SparkScan.java (hunk 15 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/storage/SparkScan.java:22]: this class returns false always. will address this  separately in refactor https://github.com/opensearch-project/sql/issues/1820

**Review by dai-chen — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java (hunk 2 above)
Comment (rupal-bq) [reply spark/src/main/java/org/opensearch/sql/spark/functions/scan/SparkSqlFunctionTableScanOperator.java:52]: created https://github.com/opensearch-project/sql/issues/1820 to address refactor. will address it in follow up PR.

**Review by penghuo — APPROVED** [review]
